src/pipeline_v3/utils/security.py

```
@@ -0,0 +1,462 @@
+        # Check for localhost
+        if hostname in ["localhost", "127.0.0.1", "::1", "0.0.0.0"]:  # noqa: S104
+            if not allow_localhost:
+                raise SecurityError(f"Localhost URLs not allowed: {url}")
+
+        # Try to resolve to IP and check if it's private
+        try:
+            # Check if hostname is already an IP
+            ip = ipaddress.ip_address(hostname)
+        except ValueError:
+            # It's a domain name, try to resolve it
+            import socket
```

Pylint (3.3.7)

[convention] Import outside toplevel (socket)

(C0415)